File: Assets/src/model/enemies/HydraHeart.cs
using System;
using System.Collections.Generic;
using System.Linq;
using UnityEngine;

[System.Serializable]
[ObjectInfo(description: "Every four turns, spawns a Hydra Head (max 12) within range 4.\nOn death, all Hydra Heads die as well.\nDoes not move or attack.", flavorText: "Thick veins writhe underneath this pulsating white mass, connecting it to an ever growing network of Heads.")]
public class HydraHeart : AIActor, IBaseActionModifier {
  public static int spawnRange = 4;
  public static bool IsTarget(Body b) {
    if (b is Player p && p.isCamouflaged) {
      return false;
    }
    return !(b is HydraHead) && !(b is HydraHeart);
  }
  private List<HydraHead> heads = new List<HydraHead>();
  public HydraHeart(Vector2Int pos) : base(pos) {
    faction = Faction.Enemy;
    hp = baseMaxHp = 9;
    // inventory.AddItem(new ItemHydraPearl());
  }

  private bool needsWait = true;
  protected override ActorTask GetNextTask() {
    if (needsWait) {
      needsWait = false;
      return new WaitTask(this, 2);
    } else {
      // update head count
      heads.RemoveAll(h => h.IsDead);
      if (heads.Count() < 12) {
        needsWait = true;
        return new TelegraphedTask(this, 1, new GenericBaseAction(this, SpawnHydraHead));
      } else {
        return new WaitTask(this, 1);
      }
    }
  }

  void SpawnHydraHead() {
    Tile spawnTile = null;

    var nearestEnemy = floor.ActorsInCircle(pos, spawnRange).Where(IsTarget).OrderBy((a) => a.DistanceTo(pos)).FirstOrDefault();
    if (nearestEnemy != null) {
      spawnTile = Util.RandomPick(floor.GetAdjacentTiles(nearestEnemy.pos).Where(CanSpawnHydraHeadAt));
    }

    if (spawnTile == null) {
      spawnTile = Util.RandomPick(floor.EnumerateCircle(base.pos, spawnRange).Select(p => floor.tiles[p]).Where(CanSpawnHydraHeadAt));
    }

    if (spawnTile != null) {
      var head = new HydraHead(spawnTile.pos);
      head.ClearTasks();
      if (this.faction == Faction.Ally) {
        head.SetAI(new CharmAI(head));
      }
      floor.Put(head);
      heads.Add(head);
    }
  }

  // make sure the hydra heart can see the tile
  private bool CanSpawnHydraHeadAt(Tile t) => t.CanBeOccupied() && floor.TestVisibility(pos, t.pos);

  public override void HandleDeath(Entity source) {
    base.HandleDeath(source);
    GameModel.main.EnqueueEvent(() => {
      foreach (var head in heads) {
        head.Kill(source);
      }
    });
  }

  internal override (int, int) BaseAttackDamage() => (0, 0);

  public BaseAction Modify(BaseAction input) {
    // cannot move
    if (input.Type == ActionType.MOVE || input.Type == ActionType.ATTACK) {
      return new WaitBaseAction(this);
    }
    return input;
  }
}

[Serializable]
[ObjectInfo("hydra-pearl")]
public class ItemHydraPearl : Item, IEdible {
  public void Eat(Actor a) {
    a.Heal(2);
    a.statuses.Add(new ConfusedStatus(5));
    Destroy();
  }

  internal override string GetStats() => "Eat to heal 2 HP and become confused for 5 turns.";
}

[System.Serializable]
[ObjectInfo(description: "Attacks anything adjacent to it.\nStationary.", flavorText: "A fleshy tube with a gaping jaw at the end, grasping at any food nearby.")]
public class HydraHead : AIActor, IBaseActionModifier, INoTurnDelay {
  public static new ActionCosts StaticActionCosts = new ActionCosts(Actor.StaticActionCosts) {
    [ActionType.ATTACK] = 2f,
  };

  protected override ActionCosts actionCosts => StaticActionCosts;
  public HydraHead(Vector2Int pos) : base(pos) {
    faction = Faction.Enemy;
    hp = baseMaxHp = 3;
  }

  internal override (int, int) BaseAttackDamage() => (1, 1);

  protected override ActorTask GetNextTask() {
    var potentialTargets = floor
      .AdjacentBodies(pos)
      .Where(HydraHeart.IsTarget);
    if (potentialTargets.Any()) {
      var target = Util.RandomPick(potentialTargets);
      return new AttackTask(this, target);
    } else {
      return new WaitTask(this, 1);
    }
  }

  public BaseAction Modify(BaseAction input) {
    // cannot move
    if (input.Type == ActionType.MOVE) {
      return new WaitBaseAction(this);
    }
    return input;
  }
}

File: web/src/model/enemies/HydraHeart.ts
import { AIActor } from './AIActor';
import { Body } from '../Body';
import { ActorTask } from '../ActorTask';
import { WaitTask } from '../tasks/WaitTask';
import { TelegraphedTask } from '../tasks/TelegraphedTask';
import { AttackGroundTask } from '../tasks/AttackGroundTask';
import { GenericBaseAction, WaitBaseAction, ActionCosts } from '../BaseAction';
import { BASE_ACTION_MOD, type IBaseActionModifier } from '../../core/Modifiers';
import { ActionType, Faction } from '../../core/types';
import type { IDeathHandler } from '../../core/types';
import { Vector2Int } from '../../core/Vector2Int';
import { GameModelRef } from '../GameModelRef';
import { MyRandom } from '../../core/MyRandom';
import { SurprisedStatus } from '../tasks/SleepTask';
import { entityRegistry } from '../../generator/entityRegistry';
import type { Entity } from '../Entity';
import type { Tile } from '../Tile';

const DEATH_HANDLER = Symbol.for('IDeathHandler');

/**
 * Spawns HydraHeads every 4 turns within range 3. Max 6 heads.
 * Cannot move or attack. On death, kills all heads.
 * Port of C# HydraHeart.cs.
 */
export class HydraHeart extends AIActor implements IBaseActionModifier, IDeathHandler {
  readonly [BASE_ACTION_MOD] = true as const;
  readonly [DEATH_HANDLER] = true;
  override get isStationary() { return true; }

  static readonly spawnRange = 3;

  private heads: HydraHead[] = [];
  private needsWait = false;

  static isTarget(b: Body): boolean {
    return !(b instanceof HydraHead) && !(b instanceof HydraHeart) && !b.isDead;
  }

  constructor(pos: Vector2Int) {
    super(pos);
    this.faction = Faction.Enemy;
    this._hp = this._baseMaxHp = 5;
  }

  baseAttackDamage(): [number, number] {
    return [0, 0];
  }

  protected getNextTask(): ActorTask {
    if (this.needsWait) {
      this.needsWait = false;
      return new WaitTask(this, 2);
    }
    // Clean up dead heads
    this.heads = this.heads.filter(h => !h.isDead);
    if (this.heads.length < 5) {
      this.needsWait = true;
      return new TelegraphedTask(this, 1, new GenericBaseAction(this, () => this.spawnHydraHead()));
    }
    return new WaitTask(this, 1);
  }

  private spawnHydraHead(): void {
    const floor = this.floor;
    if (!floor) return;

    let spawnTile: Tile | null = null;

    // Find nearest target in range
    const bodiesInRange = floor.bodiesInCircle(this.pos, HydraHeart.spawnRange)
      .filter((b): b is Body => b instanceof Body && HydraHeart.isTarget(b));
    bodiesInRange.sort((a, b) => a.distanceTo(this.pos) - b.distanceTo(this.pos));
    const nearestEnemy = bodiesInRange[0] ?? null;

    if (nearestEnemy) {
      const adjacentTiles = floor.getAdjacentTiles(nearestEnemy.pos)
        .filter(t => this.canSpawnHydraHeadAt(t));
      if (adjacentTiles.length > 0) {
        spawnTile = MyRandom.Pick(adjacentTiles);
      }
    }

    if (!spawnTile) {
      const candidates: Tile[] = [];
      for (const p of floor.enumerateCircle(this.pos, HydraHeart.spawnRange)) {
        const t = floor.tiles.get(p);
        if (t && this.canSpawnHydraHeadAt(t)) {
          candidates.push(t);
        }
      }
      if (candidates.length > 0) {
        spawnTile = MyRandom.Pick(candidates);
      }
    }

    if (spawnTile) {
      const head = new HydraHead(spawnTile.pos);
      head.clearTasks();
      head.statuses.add(new SurprisedStatus());
      floor.put(head);
      this.heads.push(head);
    }
  }

  private canSpawnHydraHeadAt(t: Tile): boolean {
    return t.canBeOccupied() && this.floor!.testVisibility(this.pos, t.pos);
  }

  handleDeath(source: Entity | null): void {
    const heads = [...this.heads];
    GameModelRef.main.enqueuEvent(() => {
      for (const head of heads) {
        if (!head.isDead) {
          head.kill(source ?? this);
        }
      }
    });
  }

  modify(input: any): any {
    if (input != null && typeof input === 'object' && 'type' in input) {
      const type = input.type;
      if (type === ActionType.MOVE || type === ActionType.ATTACK) {
        return new WaitBaseAction(this);
      }
    }
    return input;
  }
}

/**
 * Stationary head that attacks random adjacent non-Hydra targets.
 * Attack cost = 3. Cannot move.
 * Port of C# HydraHead.cs.
 */
export class HydraHead extends AIActor implements IBaseActionModifier {
  readonly [BASE_ACTION_MOD] = true as const;
  override get isStationary() { return true; }

  get turnPriority(): number {
    return this.task?.constructor.name === 'AttackGroundTask' ? 90 : super.turnPriority;
  }

  protected get actionCosts(): ActionCosts {
    return new ActionCosts([
      [ActionType.ATTACK, 1],
      [ActionType.GENERIC, 1],
      [ActionType.MOVE, 1],
      [ActionType.WAIT, 1],
    ]);
  }

  constructor(pos: Vector2Int) {
    super(pos);
    this.faction = Faction.Enemy;
    this._hp = this._baseMaxHp = 1;
  }

  baseAttackDamage(): [number, number] {
    return [1, 1];
  }

  protected getNextTask(): ActorTask {
    const floor = this.floor;
    if (!floor) return new WaitTask(this, 1);

    const targets = floor.adjacentBodies(this.pos)
      .filter((b): b is Body => b instanceof Body && HydraHeart.isTarget(b));

    if (targets.length > 0) {
      const target = MyRandom.Pick(targets);
      return new AttackGroundTask(this, target.pos, 1);
    }
    return new WaitTask(this, 1);
  }

  modify(input: any): any {
    if (input != null && typeof input === 'object' && 'type' in input) {
      if (input.type === ActionType.MOVE) {
        return new WaitBaseAction(this);
      }
    }
    return input;
  }
}

entityRegistry.register('HydraHeart', HydraHeart);
entityRegistry.register('HydraHead', HydraHead);
